### PR TITLE
synchronize clusterIP CIDR from serviceCIDR to support k8s 1.29

### PR DESCRIPTION
**Pull request by cyclinder**
it could be sync service cidr to spidercoordinator from serviceCIDR in k8s 1.29

#### What type of PR is this?

- release/feature

**What this PR does / why we need it**:

Refer to https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/1880-multiple-service-cidrs/README.md, kubernetes can allocate service ips from serviceCIDR resources in 1.29. 

```

➜  ~ kubectl get servicecidr kubernetes -o yaml
apiVersion: networking.k8s.io/v1alpha1
kind: ServiceCIDR
metadata:
  creationTimestamp: "2024-01-25T08:36:00Z"
  finalizers:
  - networking.k8s.io/service-cidr-finalizer
  name: kubernetes
  resourceVersion: "504422"
  uid: 72461b7d-fddd-409d-bdf2-83d1a2c067ca
spec:
  cidrs:
  - 10.233.0.0/18
  - fd00:10:233::/116
status:
  conditions:
  - lastTransitionTime: "2024-01-28T06:38:55Z"
    message: Kubernetes Service CIDR is ready
    reason: ""
    status: "True"
    type: Ready

➜  ~ kubectl get servicecidr kubernetes -o yaml
apiVersion: networking.k8s.io/v1alpha1
kind: ServiceCIDR
metadata:
  creationTimestamp: "2024-01-25T08:36:00Z"
  finalizers:
  - networking.k8s.io/service-cidr-finalizer
  name: kubernetes
  resourceVersion: "504422"
  uid: 72461b7d-fddd-409d-bdf2-83d1a2c067ca
spec:
  cidrs:
  - 10.233.0.0/18
  - fd00:10:233::/116
status:
  conditions:
  - lastTransitionTime: "2024-01-28T06:38:55Z"
    message: Kubernetes Service CIDR is ready
    reason: ""
    status: "True"
    type: Ready
```

This PR synchronizes the service cidr from the `serviceCIDR` resource to the spidercoordinator. If the resource cannot be found or there are any errors, then we will fall back to the default behavior (from kubeadm-config and kube-controller-manager).

```
➜  test git:(3e40ed05) ✗ kubectl get servicecidr
NAME         CIDRS                             AGE
kubernetes   10.233.0.0/18,fd00:10:233::/116   5h41m
test1        10.234.0.0/18,fd00:10:234::/116   53m
➜  test git:(3e40ed05) ✗ kubectl get spidercoordinators.spiderpool.spidernet.io default -o yaml
apiVersion: spiderpool.spidernet.io/v2beta1
kind: SpiderCoordinator
metadata:
  creationTimestamp: "2024-01-25T08:41:50Z"
  finalizers:
  - spiderpool.spidernet.io
  generation: 1
  name: default
  resourceVersion: "41645"
  uid: d1e095db-d6e8-4413-b60e-fcf31ad2bf5e
spec:
  detectGateway: false
  detectIPConflict: false
  hijackCIDR:
  - 10.244.64.0/18
  - fd00:10:244::/112
  hostRPFilter: 0
  hostRuleTable: 500
  mode: auto
  podCIDRType: auto
  podDefaultRouteNIC: ""
  podMACPrefix: ""
  tunePodRoutes: true
  txQueueLen: 0
status:
  phase: Synced
  serviceCIDR:
  - 10.234.0.0/18
  - fd00:10:234::/116
  - 10.233.0.0/18
  - fd00:10:233::/116
  
➜  test kubectl delete servicecidr test1
servicecidr.networking.k8s.io "test1" deleted

➜  test git:(3e40ed05) ✗
➜  test git:(3e40ed05) ✗ kubectl get spidercoordinators.spiderpool.spidernet.io default -o yaml
apiVersion: spiderpool.spidernet.io/v2beta1
kind: SpiderCoordinator
metadata:
  creationTimestamp: "2024-01-25T08:41:50Z"
  finalizers:
  - spiderpool.spidernet.io
  generation: 1
  name: default
  resourceVersion: "46709"
  uid: d1e095db-d6e8-4413-b60e-fcf31ad2bf5e
spec:
  detectGateway: false
  detectIPConflict: false
  hijackCIDR:
  - 10.244.64.0/18
  - fd00:10:244::/112
  hostRPFilter: 0
  hostRuleTable: 500
  mode: auto
  podCIDRType: auto
  podDefaultRouteNIC: ""
  podMACPrefix: ""
  tunePodRoutes: true
  txQueueLen: 0
status:
  phase: Synced
  serviceCIDR:
  - 10.233.0.0/18
  - fd00:10:233::/116
```

**Which issue(s) this PR fixes**:

Fixes https://github.com/spidernet-io/spiderpool/issues/2905

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by weizhoublue** [comment]
in the PR description, could you attach a yaml example of servicecidr to show what it looks like for reference

**Comment by weizhoublue** [comment]
for a new feature, it should add some explanation.
(1) doc/develop/roadmap
(2) doc/usage/coordinator.md

File: docs/develop/roadmap.md [hunk 1]
```diff
@@ -34,6 +34,9 @@
 | CNI version        | cni v1.0.0                                                                                                                             | v0.4.0        | v0.5.0       |            |
 | ifacer             | bond interface                                                                                                                         | v0.6.0        | v0.8.0       |            |
 |                    | vlan interface                                                                                                                         | v0.6.0        | v0.8.0       |            |
+| SpiderCoordinator  | Sync podCIDR for calico | v0.6.0 | v0.8.0
+|                    | Sync podCIDR for cilium | v0.6.0 | v0.8.0
+|                    | sync clusterIP CIDR from serviceCIDR to support k8s 1.29 | | v0.1.0 |
```
Comment (weizhoublue) [inline docs/develop/roadmap.md:39]: release v0.10.0 ？

File: pkg/coordinatormanager/coordinator_informer.go [hunk 2]
```diff
@@ -387,18 +389,26 @@ func (cc *CoordinatorController) fetchPodAndServerCIDR(ctx context.Context, logg
...
 
-	coordCopy.Status.ServiceCIDR = k8sServiceCIDR
+	// if we can find out the serviceCIDR(any errors), start a goroutine to sync the serviceCIDR
+	// else we use k8sServiceCIDR
+	if err = cc.fetchServiceCIDR(ctx, logger, coordCopy); err != nil {
+		logger.Sugar().Warnf("failed to fetch ServiceCIDR: %v, sync service cidr from kubeadm-config or kube-controller-manager pod", err)
+		coordCopy.Status.ServiceCIDR = k8sServiceCIDR
+	}
+
 	return coordCopy, nil
```
Comment (Icarus9913) [inline pkg/coordinatormanager/coordinator_informer.go:412]: 是否会在每次执行coordinator的`syncHandler`时候调用`fetchServiceCIDR`？ 不期望每次SpiderCoorinator的CR实例resync或变动时，都会触发`fetchServiceCIDR `启动一个新的goroutine去watch

File: pkg/coordinatormanager/service_cidr_informer.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	return c, nil
+}
+
+type serviceCIDRReconciler struct {
+	client          client.Client
+	coordinatorName string
+}
+
+func (r *serviceCIDRReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var ipPoolList networkingv1.ServiceCIDRList
```
Comment (Icarus9913) [inline pkg/coordinatormanager/service_cidr_informer.go:53]: 变量名换一下，`ipPoolList ` --> `serviceCIDRList`或`svcCIDRList`

File: pkg/coordinatormanager/service_cidr_informer.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	if err := r.client.List(ctx, &ipPoolList); err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	serviceCIDR := make([]string, 0, len(ipPoolList.Items))
+	for _, p := range ipPoolList.Items {
+		if p.DeletionTimestamp == nil {
+			serviceCIDR = append(serviceCIDR, p.Spec.CIDRs...)
+		}
+	}
```
Comment (Icarus9913) [inline pkg/coordinatormanager/service_cidr_informer.go:63]: 补一下log

File: pkg/coordinatormanager/service_cidr_informer.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	if coordinator.Status.Phase == Synced && reflect.DeepEqual(coordinator.Status.ServiceCIDR, serviceCIDR) {
+		return ctrl.Result{}, nil
+	}
+
+	origin := coordinator.DeepCopy()
+	coordinator.Status.Phase = Synced
+	coordinator.Status.ServiceCIDR = serviceCIDR
+	if err := r.client.Status().Patch(ctx, &coordinator, client.MergeFrom(origin)); err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
```
Comment (Icarus9913) [inline pkg/coordinatormanager/service_cidr_informer.go:79]: 补充一下origin和新的coordinator的`serviceCIDR`区别log

File: pkg/coordinatormanager/coordinator_informer.go (hunk 2 above)
Comment (cyclinder) [reply pkg/coordinatormanager/coordinator_informer.go:413]: 目前和calico 和 cilium 处理方式一致，每次 reconcile 时先停止前一个 goroutine，再启动一个新的，避免goroutine 泄漏

File: docs/develop/roadmap.md (hunk 1 above)
Comment (cyclinder) [reply docs/develop/roadmap.md]: done

File: pkg/coordinatormanager/service_cidr_informer.go (hunk 3 above)
Comment (cyclinder) [reply pkg/coordinatormanager/service_cidr_informer.go]: done

**Comment by ty-dc** [comment]
该 PR 似乎会导致 k8s 矩阵（1.22、1.23、1.24等版本）失败 ？是否只能在 1.29 版本的 K8S 上运行？

**Comment by cyclinder** [comment]
> 该 PR 似乎会导致 k8s 矩阵（1.22、1.23、1.24等版本）失败 ？是否只能在 1.29 版本的 K8S 上运行？

是的，e2e 只能在 1.29 以上跑

**Review by Icarus9913 — APPROVED** [review]

**Review by weizhoublue — APPROVED** [review]

**Pull request merged**